datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -433,6 +435,7 @@ pub fn array_element(args: &[ArrayRef]) -> Result<ArrayRef> {
             let indexes = as_int64_array(&args[1])?;
             general_array_element::<i64>(array, indexes)
         }
+        DataType::Null => Ok(args[0].clone()),
```

Comment:
why do we need to handle null here? Is it like handling length checking? If it is,  note that `array_element(null, )` should be an error.

Reply:
We may need functions like `pre-checking`, or `validation` for the Array function when we move to UDF-based implementation 🤔 .

Reply:
I think handling null here makes it confusing where it is nulls handled, is it handled in`coercion` step `coerce signature` or here (array_expressions). The test in array.slt only process once. We can only know it is handled correctly, but unsure which one does the job.

Reply:
> I think handling null here makes it confusing where it is nulls handled, is it handled in`coercion` step `coerce signature` or here (array_expressions). The test in array.slt only process once. We can only know it is handled correctly, but unsure which one does the job.

This is what I am considering now, in `coercion`  step could filter if the function accepts the `NULL` type while in `array_expressions` it determines what the function should return @jayzhan211 .

Reply:
> in coercion step could filter if the function accepts the NULL type while in array_expressions it determines what the function should return

Does it mean we still need to do the null handling in `array_expression`? Take `array_append` for example, if we find null, we convert it to i64(None) and process it later on. 

ideally, nulls should be handled in type coercion. And, `array_expression` can be *simple* that just focuses on processing the result and not handling type or signature-related issues.

If doing the checking again in `array_expression` is just because it is a public function, I think we should come out with a way that the checking can be somewhat reused in both places. Before that, maybe we just leave the checking in `coercion` step.

Reply:
> > in coercion step could filter if the function accepts the NULL type while in array_expressions it determines what the function should return
> 
> Does it mean we still need to do the null handling in `array_expression`? Take `array_append` for example, if we find null, we convert it to i64(None) and process it later on.
> 
> ideally, nulls should be handled in type coercion. And, `array_expression` can be _simple_ that just focuses on processing the result and not handling type or signature-related issues.
> 

It would be ideal if we could handle null in one place. However, array functions have different results for `NULL`. For instance, `array_append(null, 1)` returns an error, whereas `array_has(null, 1)` returns null.

Reply:
Maybe we can just ignore nulls as first argument since we are not coerce them to any other type. Unlike, array_append(arr, null) as second argument, in this case we will convert to other primitives.

Reply:
I think that we should only focus on *type coercion* in TypeCoercionRewriter. If we got null, but not casting it to other, ignore it.

Reply:
> I believe our primary focus within TypeCoercionRewriter should be strictly on type coercion. If encountering a null value that isn't being cast to another type, it should be disregarded.

I cannot agree more with that. I should focus on the type coercion, and leave `NULL` aside now. So I will remove `allow_null_coercion ` and don't check `null` now.